contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,470 @@
+            block.timestamp >= _stakingPoolInfo.epochStartTimestamp,
+            "StakingPool: too early"
+        );
+        _;
+    }
+
+    modifier stakingPoolNotFull(uint256 _deposit) {
+        require(
+            _stakingPoolInfo.totalStakedAmount + _deposit <
+                _stakingPoolInfo.maxTotalPoolStake,
+            "StakingPool: pool full"
+        );
+        _;
+    }
+
+    function pause()
+        external
+        whenNotPaused
+        atLeastDaoMeepleRole(_stakingPoolInfo.daoId)
+    {
+        _pause();
+    }
+
+    function unpause()
+        external
+        whenPaused
+        atLeastDaoMeepleRole(_stakingPoolInfo.daoId)
+    {
+        _unpause();
+    }
+
+    /**
+     * @notice Only entry point for a user to deposit into the staking pool
+     *
+     * @param amount Amount of stake tokens to deposit
+     */
+    function deposit(uint256 amount)
+        external
+        whenNotPaused
+        stakingPeriodNotStarted
+        nonReentrant
+        stakingPoolNotFull(amount)
+    {
+        require(
+            amount >= _stakingPoolInfo.minimumContribution,
+            "StakingPool: min contribution"
+        );
+
+        User storage user = _users[_msgSender()];
+        StakingPoolLib.Data storage _info = _stakingPoolInfo;
+
+        user.depositAmount += uint128(amount);
+        _info.totalStakedAmount += uint128(amount);
+
+        // calculate/update rewards
+        for (uint256 i = 0; i < _info.rewardTokens.length; i++) {
+            if (_info.poolType == StakingPoolLib.StakingPoolType.FLOATING) {
+                // floating: update the global rewards ratio for each reward token
+                _info
+                    .rewardTokens[i]
+                    .rewardAmountRatio = _computeFloatingRewardsPerShare(
+                    _info.rewardTokens[i].totalTokenRewardsAvailable,
+                    _info.totalStakedAmount
+                );
+            } else {
+                // fixed: set the reward amount per user now
+                user.rewardAmounts[i] += uint128(
+                    (amount * _info.rewardTokens[i].rewardAmountRatio)
+                );
+            }
+        }
+
+        emit Deposit(_msgSender(), amount);
+
+        require(
+            _info.stakeToken.transferFrom(_msgSender(), address(this), amount),
+            "StakingPool: failed to transfer"
+        );
+    }
+
+    /**
+     * @notice Withdraw both stake and reward tokens when the stake period is complete
+     */
+    function withdraw()
+        external
+        stakingPeriodComplete
+        rewardsAvailable
+        nonReentrant
+    {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+
+        StakingPoolLib.Data storage _info = _stakingPoolInfo;
+
+        //slither-disable-next-line reentrancy-events
+        _transferStake(user.depositAmount, IERC20(_info.stakeToken));
```

Comment:
nonreentry modifier should make this ok at the cost of gas consumption?

Reply:
There's an opportunity for code reuse.

As you've got individual functions of stake and reward withdrawal already, if you refactor the core logic into helpers, it'll reduce the size of the top level function & should make them easier to grok.

Reply:
non-reentry modifier will be unavoidable for this case of multiple external contract calls 🤷

Reply:
I added a function: 
`    function _transferToken(uint256 amount, IERC20 token) private {
        require(
            token.transfer(_msgSender(), amount),
            "StakingPool: tx failed"
        );
    }`
    but feel we are losing info here. Is this what you meant?

Reply:
I mean add private function to get rid of the comments